rfcs/56-better-reflect.md

```diff
@@ -0,0 +1,261 @@
+    match self { /* etc. */ }
+  }
+  pub fn can_sidecast(&self, registry: &TypeRegistry) -> Result<(), SidecastError> {}
+  pub fn sidecast(self, registry: &TypeRegistry)
+    -> Result<Box<dyn Reflect>, SidecastError> {
+    let type_name = self.type_name();
+    let registration = registry.get_with_name(type_name).ok_or(SidecastError::Unregistered)?;
+    Ok(registration.constructor(self)?)
+  }
+  pub fn sidecast_type<T: Reflect + Typed>(self) -> Result<Box<dyn Reflect>, SidecastError> {
+    let registration = TypeRegistration::of::<T>();
+    Ok(registration.constructor(self)?)
+  }
+}
+```
+ 
+`sidecast` requires introducing a new field to `TypeRegistration`:
+```rust
+pub struct TypeRegistration {
+    short_name: String,
+    // new: constructor
+    constructor: fn(ReflectProxy) -> Result<Box<dyn Reflect>, ConstructorError>,
+    data: HashMap<TypeId, Box<dyn TypeData>>,
+    type_info: TypeInfo,
+}
+impl TypeRegistration {
+  pub fn constructor(&self, proxy: ReflectProxy) ->  Result<Box<dyn Reflect>, ConstructorError> {
+    (self.constructor)(proxy)
+  }
+  // ...
+}
+```
+
+`constructor` would simply construct the `T` and wrap it into a `Box` before
+returning it.
+
+`constructor` can be derived similarly to `TypeInfo`. It will however
+require additional consuming methods on `Dynamic***` structs in order to be
+able to move their `Box<dyn Reflect>` fields into the constructed data
+structure.
+
+## Drawbacks
+
+- Users cannot define their own proxy type, they must use the `Dynamic***`
+  types.
+- `Dynamic***` do not implement the `Reflect` subtraits anymore.
+- There is some added complexity (although it fixes what I believe to be a
+  major limitation of the `bevy_reflect` API)
+- `Reflect::apply` now only accepts a `ReflectProxy` (see next section)
+
+## Rationale and alternatives
+
+- Divide `Reflect` in two distinct traits, one with the underlying-sensible
+  method, and one with the other methods. (see next points)
+- An earlier version of this RFC proposed `ReflectProxy` as a **trait**,
+  however, I reformulated it as an enum when considering how to implement it.
+  (How would you define the `TypeRegistration::constructor` field if passed a
+  `Box<dyn ReflectProxy>`? You'd need something like `ReflectProxyRef`?)
+
+I believe this change is needed, since `Reflect` duplicity is a large footgun.
+
+I don't believe that having `Dynamic***` implement `Reflect` makes sense, I
+think it introduces much more issues than it solves. The only use-case for
+`Dynamic***` being `Reflect` are the `reflect_mut`, `reflect_ref` and `apply`
+methods. However, those could as well be implemented outside of the `Reflect`
+trait, as mentioned in the "Future possibilities" section.
+
+## Unresolved questions
+
+- How does this interact with `#[reflect(ignore)]`?
+- Is a `trait ReflectProxy` more desirable than an `enum` and how to implement
+  it?
+- Should `Dynamic***` inner values be `Box<ReflectProxy>` or `Box<dyn Reflect>`?
```

Comment:
I think it would have to be `Box<ReflectProxy>`, right? Since creating the proxy requires fields be recursively `clone_proxy`'d.

Reply:
The proxy would only be shallow with `Box<dyn Reflect>`. No recursion. I personally prefer `Box<dyn Reflect>` because the ways `sidecast` can fail is limited and more predictable. While if you have a deeply nested element that cannot be sidecast, it's going to be hard to attach the correct context to the error type, and it's going to be annoying to debug.

Reply:
How do you clone the value of a `Box<dyn Reflect>`, though? Assuming it's underlying type is a complex structure like a struct. Won't it need the deeply nested proxies?

Reply:
That is an issue! Calling `Reflect::proxy_clone` and then `ReflectProxy::sidecast` them would work. I think the `Box<dyn Reflect>` vs `Box<ReflectProxy>` is a question of whether we want to force evaluation at construction time, or have supporting both usecases. It seems in the bevy spirit to support both use cases, so `Box<ReflectProxy>` seems now the better option to me.